src/Token.php

```diff
@@ -34,10 +35,34 @@ final class Token
     /** @var string */
     private $value;
 
-    public function __construct(int $type, string $value)
+    /** @var self|null */
+    private $previous;
+
+    /** @var self|null */
+    private $next;
+
+    /** @var bool */
+    private $wasPrecededByWhitespace = false;
+
+    public function __construct(int $type, string $value, ?Token $previous)
     {
         $this->type  = $type;
         $this->value = $value;
+        if ($previous !== null) {
+            $previous->next = $this;
```

Comment:
This means that a `Token` is really valid only if there is no token after it, or if you just finished building that token. Combined with generators, this produces bugs, and that's why I couldn't use one for `getFilteredTokens`… this might be a bad idea. Right now the program works because calls are made carefully, but this is flawed I'm afraid. I think I am going to have to rethink all of this.

Reply:
I think I fixed it now, with exceptions that prevent rewiring or accessing `next` - related features until the are available.

Reply:
Keeping as a draft until #45 is merged